internal_decision_state_machine.go

```diff
@@ -0,0 +1,820 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cadence
+
+import (
+	"fmt"
+
+	s "go.uber.org/cadence/.gen/go/shared"
+	"go.uber.org/cadence/common"
+	"go.uber.org/cadence/common/util"
+)
+
+type (
+	decisionState int32
+	decisionType  int32
+
+	decisionID struct {
+		decisionType decisionType
+		id           string
+	}
+
+	decisionStateMachine interface {
+		getState() decisionState
+		getID() decisionID
+		isDone() bool
+		getDecision() *s.Decision // return nil if there is no decision in current state
+		cancel()
+
+		handleStartedEvent()
+		handleCancelInitiatedEvent()
+		handleCanceledEvent()
+		handleCancelFailedEvent()
+		handleCompletionEvent()
+		handleInitiationFailedEvent()
+		handleInitiatedEvent()
+
+		handleDecisionSent()
+	}
+
+	decisionStateMachineBase struct {
+		id      decisionID
+		state   decisionState
+		history []string
+	}
+
+	activityDecisionStateMachine struct {
```

Comment:
It is actually a cool idea. The functions that generate decision based on state can be part of the table as well. It is also self documenting.